kubetest/gcedind/gce_dind.go

```diff
@@ -0,0 +1,622 @@
+		return fmt.Errorf("Could not find instance '%s'", gceInstanceName)
+	}
+	return nil
+}
+
+// saveInstanceLogs collects a serial log, service logs, and docker state
+// from the GCE instance, and saves the logs in a local artifacts directory.
+func (d *Deployer) saveInstanceLogs(artifactsDir string) error {
+	log.Printf("Saving logs from GCE instance")
+
+	// Create directory for GCE instance artifacts
+	logDir := filepath.Join(artifactsDir, "gce-instance")
+	if err := d.run("mkdir -p " + logDir); err != nil {
+		return err
+	}
+
+	// Copy the serial log for the instance
+	log.Printf("Saving serial log")
+	cmd := fmt.Sprintf("gcloud compute instances get-serial-port-output --project %s --zone %s --port 1 %s", d.project, d.zone, gceInstanceName)
+	if err := d.execCmdSaveLog(d.localExecCmder, cmd, logDir, serialLogFile); err != nil {
+		return err
+	}
+
+	// Save docker state
+	for _, dockerCommand := range dockerCommands {
+		if err := d.execCmdSaveLog(d.localExecCmder, dockerCommand.cmd, logDir, dockerCommand.logFile); err != nil {
+			return err
+		}
+	}
+
+	// Copy kernel and service logs from the instance
+	instanceExecCmder := d.newInstanceExecCmder()
+	if err := d.execCmdSaveLog(instanceExecCmder, "journalctl --output=short-precise -k", logDir, kernelLogFile); err != nil {
+		return err
+	}
+	for _, svc := range instanceServices {
+		cmd := fmt.Sprintf("journalctl --output=cat -u %s.service", svc)
+		logFile := fmt.Sprintf("%s.log", svc)
+		if err := d.execCmdSaveLog(instanceExecCmder, cmd, logDir, logFile); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// saveMasterNodeLogs collects docker state, service logs, and Kubernetes
+// system pod logs from all master node containers on the GCE instance,
+// and saves the logs in a local artifacts directory.
+func (d *Deployer) saveMasterNodeLogs(artifactsDir string) error {
+	masters, err := d.detectNodeContainers(kubeMasterPrefix)
+	if err != nil {
+		return err
+	}
+	for _, master := range masters {
+		if err := d.saveNodeLogs(master, artifactsDir, systemdServices, masterKubeLogs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// saveWorkerNodeLogs collects docker state, service logs, and Kubernetes
+// system pod logs from all worker node containers on the GCE instance,
+// and saves the logs in a local artifacts directory.
+func (d *Deployer) saveWorkerNodeLogs(artifactsDir string) error {
+	nodes, err := d.detectNodeContainers(kubeNodePrefix)
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes {
+		if err := d.saveNodeLogs(node, artifactsDir, systemdServices, nodeKubeLogs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// detectNodeContainers creates a list of names for either all master or all
+// worker node containers. It does this by running 'docker ps' and searching
+// for container names that begin with a specified name prefix.
+func (d *Deployer) detectNodeContainers(namePrefix string) ([]string, error) {
```

Comment:
Maybe you meant listing nodes? Unfortunately, listing pods won't work in this case since the master/worker node containers are running as "plain" (non-k8s) docker containers on the GCE instance. Listing nodes via kubectl would work, and might be safer, although the downside would be that the node-listing approach couldn't be used if we ever enhance the log collection to include gathering stuff from the tayga and bind9 (non-k8s) containers. Since we're not collecting tayga/bind9 info yet, I could switch this to use kubectl listing of nodes for now.

Reply:
I'm still not sold on making the tayga and bind9 tests blocking CI, because it creates an internet dependency, and it seems like it's testing your configuration more than kubernetes.

That said, why would it not be able to pull out tayga and bind9 logs? Are you running them in the "top" level of docker?

Reply:
There is one aspect of Kubernetes for which the external ping tests do add *some* value: When pinging from a pod to an external address, kube-proxy/ip6tables needs to masquerade (SNAT) the pod addresses, since the pod subnet is using private/ULA space (pod addresses aren't meaningful externally). Granted, the SNAT'd addresses will only exist on the link between the kube node and the tayga container, since tayga will swap in an IPv4 source address through NAT64.
Yes, tayga and bind9 are running on the "top" or "outer" docker layer (along with the master and worker nodes).